server/src/main/java/org/apache/cloudstack/consoleproxy/ConsoleAccessManagerImpl.java

```diff
@@ -0,0 +1,489 @@
+    @Inject
+    private UserVmDetailsDao userVmDetailsDao;
+    @Inject
+    private KeysManager keysManager;
+    @Inject
+    private AgentManager agentManager;
+    @Inject
+    private ConsoleProxyManager consoleProxyManager;
+
+    private static KeysManager secretKeysManager;
+    private final Gson gson = new GsonBuilder().create();
+
+    public static final Logger s_logger = Logger.getLogger(ConsoleAccessManagerImpl.class.getName());
+
+    private static Set<String> allowedSessions;
+
+    @Override
+    public boolean configure(String name, Map<String, Object> params) throws ConfigurationException {
+        ConsoleAccessManagerImpl.secretKeysManager = keysManager;
+        ConsoleAccessManagerImpl.allowedSessions = new HashSet<>();
+        return super.configure(name, params);
+    }
+
+    @Override
+    public ConsoleEndpoint generateConsoleEndpoint(Long vmId, String extraSecurityToken, String clientAddress) {
+        try {
+            if (accountManager == null || virtualMachineManager == null || managementServer == null) {
+                return new ConsoleEndpoint(false, null,"Console service is not ready");
+            }
+
+            if (keysManager.getHashKey() == null) {
+                String msg = "Console access denied. Ticket service is not ready yet";
+                s_logger.debug(msg);
+                return new ConsoleEndpoint(false, null, msg);
+            }
+
+            Account account = CallContext.current().getCallingAccount();
+
+            // Do a sanity check here to make sure the user hasn't already been deleted
+            if (account == null) {
+                s_logger.debug("Invalid user/account, reject console access");
+                return new ConsoleEndpoint(false, null,"Access denied. Invalid or inconsistent account is found");
+            }
+
+            VirtualMachine vm = entityManager.findById(VirtualMachine.class, vmId);
+            if (vm == null) {
+                s_logger.info("Invalid console servlet command parameter: " + vmId);
+                return new ConsoleEndpoint(false, null, "Cannot find VM with ID " + vmId);
+            }
+
+            if (!checkSessionPermission(vm, account)) {
+                return new ConsoleEndpoint(false, null, "Permission denied");
+            }
+
+            if (BooleanUtils.isTrue(ConsoleAccessManager.ConsoleProxyExtraSecurityValidationEnabled.value()) &&
+                StringUtils.isBlank(extraSecurityToken)) {
+                String errorMsg = "Extra security validation is enabled but the extra token is missing";
+                s_logger.error(errorMsg);
+                return new ConsoleEndpoint(false, errorMsg);
+            }
+
+            String sessionUuid = UUID.randomUUID().toString();
+            return generateAccessEndpoint(vmId, sessionUuid, extraSecurityToken, clientAddress);
+        } catch (Exception e) {
+            s_logger.error("Unexepected exception in ConsoleAccessManager", e);
```

Comment:
We could print the some context info here, like the `vmId` or the `clientAddress`.